qpdf-convert-server.py

```diff
@@ -0,0 +1,204 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+from collections import namedtuple
+import logging
+import os
+import subprocess
+import sys
+from tempfile import NamedTemporaryFile
+
+logging.basicConfig(format='%(message)s', stream=sys.stderr)
+
+
+###############################
+#         Utilities
+###############################
+
+def info(msg, suffix=None):
+    '''Qrexec wrapper for displaying information
+
+    `suffix` is typically only ever used when `msg` needs to overwrite
+    the line of the previous message (so as to imitate an updating
+    line). This is done by setting `suffix` to '\r'.
+    '''
+    print(msg, end=suffix, flush=True, file=sys.stderr)
+
+def die(msg):
+    '''Qrexec wrapper for displaying error messages'''
+    logging.error(msg)
+    sys.exit(1)
+
+def send(data):
+    '''Qrexec wrapper for sending text data to the client's STDOUT'''
+    print(data, flush=True)
+
+def send_b(data):
+    '''Qrexec wrapper for sending binary data to the client's STDOUT'''
+    sys.stdout.buffer.write(data)
+    sys.stdout.buffer.flush()
+
+def recv_b(size=None):
+    '''Qrexec wrapper for receiving data from a client'''
+    untrusted_data = sys.stdin.buffer.read(size)
+    return untrusted_data
+
+def recvline_b():
+    '''Qrexec wrapper for receiving a line of data from a client'''
+    untrusted_data = sys.stdin.buffer.readline()
+    return untrusted_data
+
+
+###############################
+#        Image-related
+###############################
+
+def send_img_dimensions(png_path):
+    cmd_width = ['identify', '-format', '%w', png_path]
+    cmd_height = ['identify', '-format', '%h', png_path]
+
+    try:
+        untrusted_width = subprocess.run(cmd_width, capture_output=True,
+                                         check=True).stdout.decode()
+        untrusted_height = subprocess.run(cmd_height, capture_output=True,
+                                         check=True).stdout.decode()
+    except subprocess.CalledProcessError:
+        die("Failed to gather dimensions... Aborting")
+
+    send(f'{untrusted_width} {untrusted_height}')
+
+def send_rgb_file(rgb_path):
+    with open(rgb_path, 'rb') as f:
+        data = f.read()
+        send_b(data)
+
+def pdf_to_png(pagenum, pdf_path, png_path):
+    png_filename = os.path.splitext(png_path)[0]
+    cmd = ['pdftocairo', pdf_path, '-png', '-f', str(pagenum), '-l',
+           str(pagenum), '-singlefile', png_filename]
+
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        cmd = ['convert', pdf_path, f'png:{png_path}']
```

Comment:
This will work only for single-page PDF files. But I see original version also had this issue, so it can stay as it is and be fixed separately.

Reply:
I actually ran this a couple days ago on a 20-page PDF and it did all the pages. But yeah, I just removed this in my local version.